tests/jobs/localisation/example_case/scripts/common_functions.py

```diff
@@ -0,0 +1,916 @@
+    """
+
+    directory: str = "observations"
+    file_name: str = "observations.obs"
+    data_dir: str = "obs_data"
+    reference_param_file: str = "init_files/ObsField"
+    reference_field_name: str = "ObsField"
+    rel_error: float = 0.10
+    min_abs_error: float = 0.01
+    # selected_grid_cells: Tuple[Tuple[int]] = ((5, 10, 1), (10, 5, 1))
+    selected_grid_cells: Tuple[Tuple[int]] = (8, 12, 1)
+
+
+@dataclass
+class Localisation:
+    """
+    Specify settings for the localisation config file.
+    """
+
+    method: str = "gaussian"
+    # method: str = "constant"
+
+
+@dataclass
+class Optional:
+    """
+    Specify if some optional files should be
+    written or not (for QC purpose).
+    """
+
+    write_obs_pred_diff_field_file: bool = True
+
+
+@dataclass
+class Settings:
+    """
+    Settings for the test case
+    """
+
+    grid_size: GridSize = GridSize()
+    field: Field = Field()
+    response: Response = Response()
+    observation: Observation = Observation()
+    localisation: Localisation = Localisation()
+    optional: Optional = Optional()
+
+
+settings = Settings()
+
+
+def read_config_file(config_file_name=None):
+    # Modify default settings  if config_file exists
+    if config_file_name:
+        if os.path.exists(config_file_name):
+            with open(config_file_name, "r", encoding="utf-8") as yml_file:
+                settings_yml = yaml.safe_load(yml_file)
+
+            update_settings(settings_yml)
+
+
+def update_key(key, default_value, settings_dict, parent_key=None):
+    value = settings_dict.get(key, default_value)
+    if value != default_value:
+        if parent_key:
+            print(f"Changed settings parameter for {key} under {parent_key} : {value} ")
+        else:
+            print(f"Changed settings parameter for {key} : {value} ")
+    return value
+
+
+def update_settings(config_dict: dict):
+    # pylint: disable=too-many-branches, too-many-statements
+    settings_dict = config_dict["settings"]
+
+    valid_keys = [
+        "grid_size",
+        "field",
+        "response",
+        "observation",
+        "localisation",
+        "optional",
+    ]
+    for key in settings_dict:
+        if key not in valid_keys:
+            raise KeyError(f"Unknown keyword {key} in 'settings' ")
+
+    key = "grid_size"
+    grid_size_dict = settings_dict[key] if key in settings_dict else None
+    valid_keys = [
+        "xsize",
+        "ysize",
+        "zsize",
+        "polygon_file",
+        "use_eclipse_grid_index_origo",
+    ]
+    if grid_size_dict:
+        err_msg = []
+        for sub_key in grid_size_dict:
+            if sub_key not in valid_keys:
+                err_msg.append(f"  {sub_key}")
+        if len(err_msg) > 0:
+            print(f"Unknown keywords in config file under keyword {key}: ")
+            for text in err_msg:
+                print(text)
+            raise KeyError("Unknown keywords")
+
+        grid_size_object = settings.grid_size
```

Comment:
I find this function a bit hard to read.
If I understand this correctly, it modifies parts of the global `settings` object.
I think we should try splitting it up somehow such that it does not modify global state.

Reply:
The purpose is to modify hardcoded default settings with settings from the config file.
The settings object defined by settings = Settings() is now a global object used everywhere where necessary.
I can maybe let it be instantiated in the main function instead just after reading the yaml file and update it by the modifed settings from the yaml file and let the read_config_file return the settings object and use it only in the main function. I will try that, but it requires that all functions depending on settings will need a bunch of parameters. But as I understood that was preferred over a single global object containing settings.

and call the function update_setting

Reply:
Is handled in another way now in test_full.py The example config file is still read by the forward model, but generated automatically to avoid manual editing and errors.